File: de.fu_berlin.inf.dpp.ui/src/de/fu_berlin/inf/dpp/ui/renderer/Renderer.java
package de.fu_berlin.inf.dpp.ui.renderer;

import de.fu_berlin.inf.ag_se.browser.extensions.IJQueryBrowser;
import de.fu_berlin.inf.dpp.ui.manager.ContactListManager;
import de.fu_berlin.inf.dpp.ui.model.ContactList;

/**
 * Implementations of this interface can transfer their current state to
 * a browser by calling Javascript functions.
 *
 * As the responsible browser instance may change, all renderers manage a reference
 * to the current browser and provide methods for its replacement.
 *
 * As those browsers may be temporary removed and re-created, the renderers
 * have to deal with null values for the browser property.
 *
 * @JTourBusStop 3, Extending the HTML GUI, Calling Javascript from Java:
 *
 *              Each renderer class is used to transfer parts of the application
 *              state to the browser for rendering.
 *
 *              They may store state themselves, as {@link ContactListRenderer}
 *              does.
 *
 *              The renderer classes call directly into the browser so that
 *              they have to know the currently active browser.
 *
 *              There are two ways to get the state from the Saros:
 *              1. use a listener if supported. See {@link ContactListManager}
 *                 for an example.
 *              2. query the state from the core directly; the {@link AccountRenderer}
 *                 does that.
 *
 *              For the management of GUI state create custom GUI model class
 *              in the model package, like {@link ContactList}.
 *              Those classes should be converted to JSON strings with the GSON
 *              library in the renderer classes.
 */
public abstract class Renderer {

    protected IJQueryBrowser browser;

    /**
     * Displays the current state managed by the renderer in the browser.
     */
    public abstract void render();

    /**
     * Sets the current browser instance, any reference to a previous
     * browser is replaced.
     *
     * @param browser the browser to be set
     */
    public synchronized void setBrowser(IJQueryBrowser browser) {
        this.browser = browser;
        render();
    }

    /**
     * Removes the set browser. This method should be called every time the browser
     * is disposed.
     */
    public synchronized void removeBrowser() {
        browser = null;
    }
}

File: de.fu_berlin.inf.dpp.ui/src/de/fu_berlin/inf/dpp/ui/renderer/Renderer.kt
package de.fu_berlin.inf.dpp.ui.renderer

import java.util.ArrayList
import de.fu_berlin.inf.ag_se.browser.extensions.IJQueryBrowser
import de.fu_berlin.inf.dpp.ui.model.State

/**
 * Implementations of this interface can transfer their current state to a list
 * of browsers by calling Javascript functions.
 *
 * As the responsible browser instances may change, all renderers manage a
 * reference to the current browsers and provide methods for its replacement.
 *
 *
 * @JTourBusStop 3, Extending the HTML GUI, Calling Javascript from Java:
 *
 * Each renderer class is used to transfer parts of the
 * application state to the each browser for rendering.
 *
 * They may store state themselves, as {@link StateRenderer} does.
 *
 * There are two ways to get the state from the Saros:
 *
 * 1. use a listener if supported. See {@link StateRenderer} for
 * an example.
 *
 * 2. query the state from the core directly the
 * {@link AccountRenderer} does that.
 *
 * For the management of GUI state create custom GUI model class
 * in the model package, like {@link State}. Those classes should
 * be converted to JSON strings with the GSON library in the
 * renderer classes.
 */
abstract class Renderer {
	private val browserList = ArrayList<IJQueryBrowser?>()
	/**
	 * Renders the current state managed by the renderer in the given browser.
	 *
	 * @param browser
	 * the browser to be rendered
	 */
	abstract fun render(browser: IJQueryBrowser?)

	/**
	 * Renders the current state managed by the renderer for each browser.
	 */
	@Synchronized
	fun render() {
		for (browser in browserList) {
			this.render(browser)
		}
	}

	/**
	 * Adds the given browser to the renderer
	 *
	 * @param browser
	 * the browser to be added
	 */
	@Synchronized
	fun addBrowser(browser: IJQueryBrowser?) {
		this.browserList.add(browser)
		render(browser)
	}

	/**
	 * Removes the given browser from the renderer. This method must be called
	 * every time the browser is disposed.
	 *
	 * @param browser
	 * the browser to be removed
	 */
	@Synchronized
	fun removeBrowser(browser: IJQueryBrowser?) {
		browserList.remove(browser)
	}
}